lib/models/finamp_models.dart

```diff
@@ -3814,3 +3716,151 @@ class RadioCacheState {
         (currentSeedItem == seedItem || queueing);
   }
 }
+
+@HiveType(typeId: 110)
+@immutable
+// Migration adapter uses hive type id 61.  We extend an empty class to prevent adapter conflicts.
+class FinampStorableQueueInfo extends FinampStorableQueueInfoLegacy {
+  const FinampStorableQueueInfo({
+    required this.currentTrackSeek,
+    required this.creation,
+    required this.sourceList,
+    required this.order,
+    required this.packedPreviousTracks,
+    required this.packedCurrentTrack,
+    required this.packedNextUp,
+    required this.packedQueue,
+    required this.sourceIndex,
+    required this.trackSourceIndexes,
+  });
+
+  factory FinampStorableQueueInfo.fromQueueInfo(FinampQueueInfo info, int? seek, FinampPlaybackOrder? order) {
+    final List<QueueItemSource> sourceList = [];
+    final List<int> sourceIndexes = [];
+
+    int addSource(QueueItemSource source) {
+      final index = sourceList.indexOf(source);
+      if (index >= 0) return index;
+      sourceList.add(source);
+      return sourceList.length - 1;
+    }
+
+    void appendTrackSource(FinampQueueItem track) {
+      sourceIndexes.add(addSource(track.source));
+    }
+
+    // Tracks must be processed in order due to appending to sourceIndexes.
+    // All sources must be added to sourceList before sourceIndexes can be bitpacked.
+    info.previousTracks.forEach(appendTrackSource);
+    if (info.currentTrack != null) appendTrackSource(info.currentTrack!);
+    info.nextUp.forEach(appendTrackSource);
+    info.queue.forEach(appendTrackSource);
+    final queueSource = addSource(info.source);
+
+    assert(sourceIndexes.length == info.trackCount);
+    assert(sourceList.isNotEmpty);
+
+    // BitBuffer throws exception attempting to write 0 bit entries, so create empty buffer manually.
+    final buffer = sourceList.length == 1
+        ? BitBuffer()
+        : BitBuffer.fromBits(sourceIndexes, bitsPerIndex: (sourceList.length - 1).bitLength);
```

Comment:
I understand packing the ID strings for each base item, but isn't compacting the indices a bit over-engineered? It's a `List<int>` with deduplicated entries already, so this should be less than a kilobyte in almost all cases?